File: src/main/java/org/example/rf/util/InputValidator.java
package org.example.rf.util;

import java.util.regex.Pattern;
import java.util.regex.Matcher;
import java.text.ParseException;
import java.text.SimpleDateFormat;

public class InputValidator {

    // Check if string is null or empty
    public static boolean isEmpty(String input) {
        return input == null || input.trim().isEmpty();
    }

    // Validate email
    public static boolean isValidEmail(String email) {
        if (isEmpty(email)) return false;
        String regex = "^[\\w.-]+@[\\w.-]+\\.[a-zA-Z]{2,6}$";
        return email.matches(regex);
    }

    // Validate phone number (e.g., 10 digits)
    public static boolean isValidPhone(String phone) {
        if (isEmpty(phone)) return false;
        String regex = "^(\\+\\d{1,3}[- ]?)?\\d{9,15}$";
        return phone.matches(regex);
    }

    // Validate integer
    public static boolean isInteger(String input) {
        try {
            Integer.parseInt(input);
            return true;
        } catch (NumberFormatException e) {
            return false;
        }
    }

    // Validate floating number
    public static boolean isFloat(String input) {
        try {
            Float.parseFloat(input);
            return true;
        } catch (NumberFormatException e) {
            return false;
        }
    }

    // Validate string length
    public static boolean isLengthBetween(String input, int min, int max) {
        if (input == null) return false;
        int length = input.length();
        return length >= min && length <= max;
    }

    // Validate date format (e.g., "yyyy-MM-dd")
    public static boolean isValidDate(String date, String format) {
        if (isEmpty(date)) return false;
        try {
            SimpleDateFormat sdf = new SimpleDateFormat(format);
            sdf.setLenient(false); // exact match
            sdf.parse(date);
            return true;
        } catch (ParseException e) {
            return false;
        }
    }

    // Validate strong password
    public static boolean isStrongPassword(String password) {
        if (isEmpty(password)) return false;
        String regex = "^(?=.*[0-9])(?=.*[a-z])(?=.*[A-Z])(?=.*[@#$%^&+=!]).{8,}$";
        return password.matches(regex);
    }

    // Validate number in range
    public static boolean isInRange(int number, int min, int max) {
        return number >= min && number <= max;
    }

    // Validate URL
    public static boolean isValidURL(String url) {
        if (isEmpty(url)) return false;
        String regex = "^(https?:\\/\\/)?[\\w.-]+\\.[a-zA-Z]{2,}(:\\d+)?(\\/.*)?$";
        return url.matches(regex);
    }

    // === Extended Validators ===

    // Check if string only contains letters and spaces (for name)
    public static boolean isAlphabetic(String input) {
        if (isEmpty(input)) return false;
        return input.matches("^[A-Za-zÀ-ỹ\\s]+$");
    }

    // Format name: "nguyễn văn a" → "Nguyễn Văn A"
    public static String formatName(String input) {
        if (isEmpty(input)) return "";
        input = input.trim().toLowerCase().replaceAll("\\s+", " ");
        String[] words = input.split(" ");
        StringBuilder sb = new StringBuilder();
        for (String word : words) {
            if (word.length() > 0) {
                sb.append(Character.toUpperCase(word.charAt(0)));
                sb.append(word.substring(1));
                sb.append(" ");
            }
        }
        return sb.toString().trim();
    }

    // Remove special characters (keep only letters, digits, space)
    public static String removeSpecialChars(String input) {
        if (isEmpty(input)) return "";
        return input.replaceAll("[^\\w\\s]", "");
    }

    // Normalize string (trim and lowercase)
    public static String normalize(String input) {
        if (isEmpty(input)) return "";
        return input.trim().toLowerCase();
    }

    // Match 2 strings exactly
    public static boolean match(String a, String b) {
        return a != null && a.equals(b);
    }

    // Check if string contains only digits
    public static boolean hasOnlyDigits(String input) {
        return input != null && input.matches("^\\d+$");
    }

}

File: src/main/webapp/js/InputValidator.js
const InputValidator = {

    // === VALIDATION RULES ===

    isEmpty: function(input) {
        return !input || input.trim() === "";
    },

    isValidEmail: function(email) {
        return /^[\w.-]+@[\w.-]+\.\w{2,}$/.test(email);
    },

    isValidPhone: function(phone) {
        return /^\+?\d{9,15}$/.test(phone);
    },

    isInteger: function(input) {
        return /^-?\d+$/.test(input);
    },

    isFloat: function(input) {
        return /^-?\d+(\.\d+)?$/.test(input);
    },

    isLengthBetween: function(input, min, max) {
        return input.length >= min && input.length <= max;
    },

    isValidDate: function(dateStr) {
        const date = new Date(dateStr);
        return !isNaN(date.getTime());
    },

    isStrongPassword: function(password) {
        return /^(?=.*[0-9])(?=.*[a-z])(?=.*[A-Z])(?=.*[@#$%^&+=!]).{8,}$/.test(password);
    },

    isInRange: function(number, min, max) {
        return number >= min && number <= max;
    },

    isValidURL: function(url) {
        return /^(https?:\/\/)?[\w.-]+\.\w{2,}(\/.*)?$/.test(url);
    },

    isAlphabetic: function(input) {
        return /^[A-Za-zÀ-ỹ\s]+$/.test(input);
    },

    // === STRING PROCESSING ===

    formatName: function(name) {
        return name
            .toLowerCase()
            .trim()
            .replace(/\s+/g, ' ')
            .split(' ')
            .map(word => word.charAt(0).toUpperCase() + word.slice(1))
            .join(' ');
    },

    removeSpecialChars: function(input) {
        return input.replace(/[^\w\s]/gi, '');
    },

    normalize: function(input) {
        return input.trim().toLowerCase();
    },

    // === AUTO FORM VALIDATION (alert từng lỗi ngay) ===

    validateFormInputs: function(formName) {
        const form = document.forms[formName];

        for (let i = 0; i < form.elements.length; i++) {
            const field = form.elements[i];
            const rule = field.getAttribute("data-validate");
            const label = field.getAttribute("data-label") || field.name || field.id || "Trường dữ liệu";

            if (!rule) continue;

            const value = field.value;

            switch (rule) {
                case "required":
                    if (this.isEmpty(value)) {
                        alert(`${label} không được để trống.`);
                        field.focus();
                        return false;
                    }
                    break;

                case "email":
                    if (!this.isValidEmail(value)) {
                        alert(`${label} không hợp lệ.`);
                        field.focus();
                        return false;
                    }
                    break;

                case "phone":
                    if (!this.isValidPhone(value)) {
                        alert(`${label} không hợp lệ.`);
                        field.focus();
                        return false;
                    }
                    break;

                case "password":
                    if (!this.isStrongPassword(value)) {
                        alert(`${label} quá yếu. Cần ít nhất 8 ký tự, gồm chữ hoa, thường, số và ký tự đặc biệt.`);
                        field.focus();
                        return false;
                    }
                    break;

                case "url":
                    if (!this.isValidURL(value)) {
                        alert(`${label} không hợp lệ.`);
                        field.focus();
                        return false;
                    }
                    break;

                case "number":
                    if (!this.isFloat(value)) {
                        alert(`${label} phải là số.`);
                        field.focus();
                        return false;
                    }
                    break;

                case "integer":
                    if (!this.isInteger(value)) {
                        alert(`${label} phải là số nguyên.`);
                        field.focus();
                        return false;
                    }
                    break;

                case "date":
                    if (!this.isValidDate(value)) {
                        alert(`${label} không phải ngày hợp lệ.`);
                        field.focus();
                        return false;
                    }
                    break;

                case "name":
                    if (this.isEmpty(value)) {
                        alert(`${label} không được để trống.`);
                        field.focus();
                        return false;
                    } else if (!this.isAlphabetic(value)) {
                        alert(`${label} chỉ được chứa chữ cái.`);
                        field.focus();
                        return false;
                    } else {
                        field.value = this.formatName(value); // format tên
                    }
                    break;

                // Bạn có thể thêm các rule khác tại đây
            }
        }

        return true; // Tất cả hợp lệ
    }
};
